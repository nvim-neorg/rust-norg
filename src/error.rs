use chumsky::error::Simple;

use crate::{stage_1::NorgToken, stage_2::NorgBlock, NorgASTFlat};

/// Represents errors that can occur during the parsing process across different stages.
#[derive(Debug)]
pub enum NorgParseError {
    Stage1(Vec<Simple<char>>),
    Stage2(Vec<Simple<NorgToken>>),
    Stage3(Vec<Simple<NorgBlock>>),
    Stage4(Vec<Simple<NorgASTFlat>>),
}

impl From<Vec<Simple<char>>> for NorgParseError {
    fn from(error: Vec<Simple<char>>) -> Self {
        NorgParseError::Stage1(error)
    }
}

impl From<Vec<Simple<NorgToken>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgToken>>) -> Self {
        NorgParseError::Stage2(error)
    }
}

impl From<Vec<Simple<NorgBlock>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgBlock>>) -> Self {
        NorgParseError::Stage3(error)
    }
}

impl From<Vec<Simple<NorgASTFlat>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgASTFlat>>) -> Self {
        NorgParseError::Stage4(error)
    }
}
