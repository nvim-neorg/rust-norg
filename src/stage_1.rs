//! This file contains the initial lexing stage, which breaks up characters into distinct tokens.

use std::fmt::Write as _;

use chumsky::prelude::*;
use chumsky::{
    text::{keyword, Character},
    Parser,
};
use serde::Serialize;
use unicode_categories::UnicodeCategories;

/// Describes an individual part of the document.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum NorgToken {
    Whitespace(u16),
    SingleNewline,
    Newlines(u16),
    Regular(char),
    Special(char),
    Escape(char),
    End(char),
    Eof,
}

impl std::fmt::Display for NorgToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::End(c) => write!(f, "{}end", c),
            Self::Eof => f.write_char('\0'),
            Self::Escape(c) => write!(f, "\\{}", c),
            Self::Newlines(count) => f.write_str(&" ".repeat(*count as usize)),
            Self::Regular(c) | Self::Special(c) => f.write_char(*c),
            Self::SingleNewline => f.write_char('\n'),
            Self::Whitespace(count) => f.write_str(&" ".repeat(*count as usize)),
        }
    }
}

impl From<NorgToken> for String {
    fn from(value: NorgToken) -> Self {
        value.to_string()
    }
}

/// A list of characters which are considered "special", i.e. for parsing of attached modifiers.
const SPECIAL_CHARS: &str = "*-~/_!%^,\"'$:@|=.#+<>()[]{}\\";

/// Parses a `.norg` document and breaks it up into tokens.
pub fn stage_1() -> impl Parser<char, Vec<NorgToken>, Error = chumsky::error::Simple<char>> {
    let ws = filter(|c: &char| c.is_inline_whitespace() || c.is_separator_space())
        .repeated()
        .at_least(1)
        .map(|content| NorgToken::Whitespace(content.len() as u16));

    // Fallback parser for any non-special character.
    let character = any().map(NorgToken::Regular);

    let parse_newline = filter(|c: &char| {
        *c == '\n' || *c == '\r' || c.is_separator_line() || c.is_separator_paragraph()
    });

    let newline = parse_newline
        .then_ignore(ws.repeated())
        .to(NorgToken::SingleNewline);

    let newlines = parse_newline
        .repeated()
        .at_least(2)
        .then_ignore(ws.repeated())
        .map(|content| NorgToken::Newlines(content.len() as u16));

    let special = one_of(SPECIAL_CHARS).map(NorgToken::Special);

    let escape = just('\\').ignore_then(any()).map(NorgToken::Escape);

    let tag_end = one_of(SPECIAL_CHARS)
        .then_ignore(keyword("end"))
        .then_ignore(choice((one_of("\n\r").rewind().map(|_| ()), end())))
        .map(NorgToken::End);

    choice((tag_end, escape, special, newlines, newline, ws, character))
        .repeated()
        .chain(end().to(NorgToken::Eof))
}
