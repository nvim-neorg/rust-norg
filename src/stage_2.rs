//! Converts a set of Norg tokens into a set of blocks.

use std::fmt::Write as _;

use chumsky::Parser;
use itertools::Itertools;
use serde::Serialize;

use crate::stage_1::NorgToken;
use chumsky::prelude::*;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum ParagraphSegmentToken {
    Text(String),
    Whitespace,
    Special(char),
    Escape(char),
}

impl std::fmt::Display for ParagraphSegmentToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Escape(c) => write!(f, "\\{}", c),
            Self::Text(str) => f.write_str(str),
            Self::Special(c) => f.write_char(*c),
            Self::Whitespace => f.write_char(' '),
        }
    }
}

impl From<ParagraphSegmentToken> for String {
    fn from(value: ParagraphSegmentToken) -> Self {
        value.to_string()
    }
}

pub type ParagraphTokenList = Vec<ParagraphSegmentToken>;

fn tokens_to_paragraph_segment(tokens: Vec<NorgToken>) -> ParagraphTokenList {
    tokens
        .into_iter()
        .peekable()
        .batching(|it| match it.next() {
            Some(NorgToken::SingleNewline) | Some(NorgToken::Whitespace(_)) => {
                Some(ParagraphSegmentToken::Whitespace)
            }
            Some(NorgToken::Special(c)) => Some(ParagraphSegmentToken::Special(c)),
            Some(NorgToken::Escape(c)) => Some(ParagraphSegmentToken::Escape(c)),
            Some(NorgToken::Regular(c)) => {
                let mut result: String = it
                    .peeking_take_while(|token| matches!(token, NorgToken::Regular(_)))
                    .map_into::<String>()
                    .collect();

                result.insert(0, c);

                Some(ParagraphSegmentToken::Text(result))
            }
            None => None,
            x => {
                unreachable!();
            }
        })
        .collect()
}

/// Represents various Norg blocks parsed from tokens.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum NorgBlock {
    /// A segment of a paragraph consisting of Norg tokens.
    ParagraphSegment(ParagraphTokenList),
    /// End of a paragraph segment.
    ParagraphSegmentEnd(ParagraphTokenList),
    /// A heading with a specified level, title, and optional extension section.
    Heading {
        level: u16,
        title: ParagraphTokenList,
        extension_section: ParagraphTokenList,
    },
    /// A nestable detached modifier with a type, level, and optional extension section.
    NestableDetachedModifier {
        modifier_type: char,
        level: u16,
        extension_section: ParagraphTokenList,
    },
    /// A rangeable detached modifier with an indication if it is ranged, type, title, and optional extension section.
    RangeableDetachedModifier {
        ranged: bool,
        modifier_type: char,
        title: ParagraphTokenList,
        extension_section: ParagraphTokenList,
    },
    /// Closing tag for a rangeable detached modifier.
    RangeableDetachedModifierClose(char),
    /// A ranged tag with a type, name, and optional parameters.
    RangedTag {
        tag_type: char,
        name: ParagraphTokenList,
        parameters: Option<Vec<ParagraphTokenList>>,
    },
    /// End of a ranged tag.
    RangedTagEnd(char),
    /// A verbatim ranged tag with a name, optional parameters, and content.
    VerbatimRangedTag {
        name: ParagraphTokenList,
        parameters: Option<Vec<ParagraphTokenList>>,
        content: Vec<NorgToken>,
    },
    /// An infirm tag with a name and optional parameters.
    InfirmTag {
        name: ParagraphTokenList,
        parameters: Option<Vec<ParagraphTokenList>>,
    },
    /// A carryover tag with a type, name, and optional parameters.
    CarryoverTag {
        tag_type: char,
        name: ParagraphTokenList,
        parameters: Option<Vec<ParagraphTokenList>>,
    },
    /// A delimiting modifier, defined by a single char `-` (weak), `=` (string), or `_` (horizontal rule)
    DelimitingModifier(char),
}

/// Defines the parser for stage 2 of the Norg parsing process, which converts tokens into blocks.
///
/// # Returns
///
/// * A parser that processes `NorgToken`s into a vector of `NorgBlock`s, which properly define
///   paragraph boundaries.
pub fn stage_2() -> impl Parser<NorgToken, Vec<NorgBlock>, Error = chumsky::error::Simple<NorgToken>>
{
    use NorgToken::*;

    let whitespace = select! { Whitespace(_) => () };

    let newlines_or_eof = select! {
        s @ SingleNewline => s,
        n @ Newlines(..) => n,
        e @ Eof => e,
    };

    let newlines_whitespace = select! {
        Newlines(_) => (),
        SingleNewline => (),
        Whitespace(_) => (),
    };

    let newlines_whitespace_or_eof = select! {
        Newlines(_) => (),
        SingleNewline => (),
        Whitespace(_) => (),
        Eof => (),
    };

    let paragraph_segment = newlines_or_eof.not().repeated().at_least(1);

    let extension_section = select! {
        SingleNewline => (),
        Newlines(_) => (),
        Eof => (),
        Special(')') => (),
    }
    .not()
    .repeated()
    .at_least(1)
    .delimited_by(just(Special('(')), just(Special(')')));

    let parameters = newlines_whitespace_or_eof
        .not()
        .repeated()
        .at_least(1)
        .separated_by(whitespace.repeated().at_least(1));

    let heading = select! {
        Special('*') => (),
    }
    .ignored()
    .repeated()
    .at_least(1)
    .map(|chars| chars.len() as u16)
    .then_ignore(whitespace.repeated().at_least(1))
    .then(extension_section.clone().or_not())
    .then(paragraph_segment)
    .then_ignore(newlines_or_eof)
    .map(|((level, extension_section), title)| NorgBlock::Heading {
        level,
        title: tokens_to_paragraph_segment(title),
        extension_section: extension_section
            .map(tokens_to_paragraph_segment)
            .unwrap_or_default(),
    })
    .labelled("heading");

    let nestable_detached_modifier = select! {
        Special(c) if c == '-' || c == '~' || c == '>' => c,
    }
    .repeated()
    .at_least(1)
    .try_map(|chars, span| {
        if chars.iter().all_equal() {
            Ok((chars[0], chars.len() as u16))
        } else {
            // Get the type of element that the user tried to create.
            let modifier_type = match chars[0] {
                '-' => "unordered list",
                '~' => "ordered list",
                '>' => "quote",
                _ => unreachable!(),
            };
            Err(Simple::custom(
                span,
                format!("
                    Expected a sequence of '{}' characters when creating {}.
                    Norg does not permit mixing of modifiers, e.g. `-~>`. Keep all your modifiers the same, e.g. `---`.
                ", chars[0], modifier_type),
            ))
        }
    })
    .then_ignore(whitespace.repeated().at_least(1))
    .then(extension_section.clone().or_not())
    .map(
        |((modifier_type, level), extension_section)| NorgBlock::NestableDetachedModifier {
            modifier_type,
            level,
            extension_section: extension_section.map(tokens_to_paragraph_segment).unwrap_or_default(),
        },
    )
    .labelled("nestabled_detached_modifier");

    let rangeable_mod = |c: char| {
        select! { Special(x) if x == c => x }
            .repeated()
            .at_least(1)
            .at_most(2)
            .map(|chars| (chars[0], chars.len() == 2))
            .then_ignore(whitespace.repeated().at_least(1))
            .then(extension_section.clone().or_not())
            .then(paragraph_segment)
            .then_ignore(newlines_or_eof)
            .map(|(((modifier_type, ranged), extension_section), title)| {
                NorgBlock::RangeableDetachedModifier {
                    modifier_type,
                    ranged,
                    title: tokens_to_paragraph_segment(title),
                    extension_section: extension_section
                        .map(tokens_to_paragraph_segment)
                        .unwrap_or_default(),
                }
            })
            .labelled("rangeable_detached_modifier")
    };

    let rangeable_mod_closer = |c: char| {
        select! { Special(x) if x == c => x }
            .repeated()
            .exactly(2)
            .ignored()
            .then_ignore(newlines_or_eof)
            .map(move |_| NorgBlock::RangeableDetachedModifierClose(c))
            .labelled("rangeable_detached_modifier_closed")
    };

    let verbatim_ranged_tag = |c: char| {
        let parse_char = select! { Special(x) if x == c => x };
        let tag_end = select! {
                End(x) if x == c => x,
        };

        let tag_parameters = select! {
            Newlines(_) => (),
            SingleNewline => (),
            Whitespace(_) => (),
            Eof => (),
            End(x) if x == c => ()
        }
        .not()
        .repeated()
        .at_least(1)
        .separated_by(whitespace.repeated().at_least(1));

        parse_char
            .ignore_then(newlines_whitespace_or_eof.not().repeated().at_least(1))
            .then(
                whitespace
                    .repeated()
                    .at_least(1)
                    .ignore_then(tag_parameters)
                    .or_not(),
            )
            .then_ignore(just(SingleNewline).or_not())
            .then_ignore(filter(|c| matches!(c, Newlines(_))).or_not())
            .then(tag_end.not().repeated().or_not())
            .then_ignore(tag_end)
            .map(
                |((name, parameters), content)| NorgBlock::VerbatimRangedTag {
                    name: tokens_to_paragraph_segment(name),
                    parameters: parameters.map(|tokens| {
                        tokens
                            .into_iter()
                            .map(tokens_to_paragraph_segment)
                            .collect()
                    }),
                    content: content.unwrap_or(vec![]),
                },
            )
    };

    let ranged_tag = |c: char| {
        let parse_char = select! { Special(x) if x == c => x };

        parse_char
            .ignore_then(newlines_whitespace_or_eof.not().repeated().at_least(1))
            .then(
                whitespace
                    .repeated()
                    .at_least(1)
                    .ignore_then(parameters)
                    .or_not(),
            )
            .then_ignore(select! {
                SingleNewline => (),
                Newlines(_) => (),
            })
            .map(move |(name, parameters)| NorgBlock::RangedTag {
                tag_type: c,
                name: tokens_to_paragraph_segment(name),
                parameters: parameters.map(|tokens| {
                    tokens
                        .into_iter()
                        .map(tokens_to_paragraph_segment)
                        .collect()
                }),
            })
    };

    let infirm_tag = {
        select! { Special('.') => '.' }
            .ignore_then(newlines_whitespace_or_eof.not().repeated().at_least(1))
            .then(
                whitespace
                    .repeated()
                    .at_least(1)
                    .ignore_then(parameters)
                    .or_not(),
            )
            .then_ignore(select! {
                SingleNewline => (),
                Newlines(_) => (),
            })
            .map(|(name, parameters)| NorgBlock::InfirmTag {
                name: tokens_to_paragraph_segment(name),
                parameters: parameters.map(|tokens| {
                    tokens
                        .into_iter()
                        .map(tokens_to_paragraph_segment)
                        .collect()
                }),
            })
    };

    let carryover_tags = {
        select! {
            Special('+') => '+',
            Special('#') => '#',
        }
        .then(newlines_whitespace_or_eof.not().repeated().at_least(1))
        .then(
            whitespace
                .repeated()
                .at_least(1)
                .ignore_then(parameters)
                .or_not(),
        )
        .then_ignore(select! {
        Newlines(_) => (),
        SingleNewline => (),
        })
        .map(|((tag_type, name), parameters)| NorgBlock::CarryoverTag {
            tag_type,
            name: tokens_to_paragraph_segment(name),
            parameters: parameters.map(|tokens| {
                tokens
                    .into_iter()
                    .map(tokens_to_paragraph_segment)
                    .collect()
            }),
        })
    };

    let tag_end = select! {
        NorgToken::End(c) => NorgBlock::RangedTagEnd(c),
    };

    let delimiting_mod = select! {
        NorgToken::Special(c @ ('-' | '=' | '_')) => c,
    }
    .repeated()
    .at_least(2)
    .then_ignore(newlines_or_eof)
    .map(|chars| NorgBlock::DelimitingModifier(chars[0]));

    choice((
        heading,
        nestable_detached_modifier,
        delimiting_mod,
        rangeable_mod('$'),
        rangeable_mod_closer('$'),
        rangeable_mod('^'),
        rangeable_mod_closer('^'),
        rangeable_mod(':'),
        rangeable_mod_closer(':'),
        verbatim_ranged_tag('@'),
        ranged_tag('|'),
        ranged_tag('='),
        infirm_tag,
        carryover_tags,
        tag_end,
        paragraph_segment
            .then(newlines_or_eof.repeated().at_least(1).rewind())
            .map(|(content, trailing)| match trailing.last().unwrap() {
                NorgToken::Eof => {
                    NorgBlock::ParagraphSegmentEnd(tokens_to_paragraph_segment(content))
                }
                NorgToken::Newlines(_) => {
                    NorgBlock::ParagraphSegmentEnd(tokens_to_paragraph_segment(content))
                }
                NorgToken::SingleNewline => NorgBlock::ParagraphSegment(
                    tokens_to_paragraph_segment(content.into_iter().chain(trailing).collect()),
                ),
                _ => unreachable!(),
            })
            .labelled("paragraph_segment"),
    ))
    .padded_by(newlines_whitespace.repeated())
    .repeated()
    .then_ignore(just(Eof))
}
