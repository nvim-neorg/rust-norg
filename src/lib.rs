use chumsky::Parser as _;
use error::NorgParseError;

use crate::stage_1::stage_1;
use crate::stage_2::stage_2;
use crate::stage_3::stage_3;

pub use crate::stage_2::ParagraphSegmentToken;
pub use crate::stage_3::{
    CarryoverTag, DetachedModifierExtension, NestableDetachedModifier, NorgASTFlat,
    RangeableDetachedModifier, RangedTag, TodoStatus,
};

mod error;
mod stage_1;
mod stage_2;
mod stage_3;

/// Parses the given input string through multiple stages to produce a flattened abstract syntax tree (AST).
///
/// # Arguments
///
/// * `input` - A string slice that holds the input to be parsed.
///
/// # Returns
///
/// * `Ok(Vec<NorgASTFlat>)` if parsing is successful.
/// * `Err(NorgParseError)` if any stage of parsing fails.
pub fn parse(input: &str) -> Result<Vec<NorgASTFlat>, NorgParseError> {
    Ok(stage_3().parse(stage_2().parse(stage_1().parse(input)?)?)?)
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;
    use itertools::Itertools;
    use proptest::{prop_oneof, proptest};

    use crate::parse;

    const TAG_NAME_REGEX: &str = r"[\w_\-\.\d]+";
    const TAG_PARAMETER_REGEX: &str = r"[^\s]+";
    const TAG_MULTI_PARAMETER_REGEX: &str = r"[^\n\r]+";

    const PARAGRAPH_REGEX: &str = r"[^[:punct:]\s][^\n\r]*";

    #[test]
    fn headings() {
        let examples: Vec<_> = [
            "* Heading",
            "********* Heading",
            "
            * Heading
              content.
            ",
            "
            ******* Heading
            ",
            "
            * Heading
            * Another heading
            ",
            "
            * Heading
            ** Subheading
            * Back to regular heading
            ",
            "
            * Heading
              sneaky content.
            ** Subheading
               more sneaky content inside.
            * Back to regular heading
            ",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn lists() {
        let examples: Vec<_> = [
            "- Test list",
            "---- Test list",
            "
                - Test list
                - Test list
                -- Test list
                -- Test list
                - Test list
                --- Test list
            ",
            "---not list",
            // "- - a list item",
            "--> not a list",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn lists_regressions() {
        [
            "- - a list item",
            "---- - a list item",
            "---- > a list item",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn ordered_lists() {
        let examples: Vec<_> = [
            "~ Test list",
            "~~~~ Test list",
            "
                ~ Test list
                ~ Test list
                ~~ Test list
                ~~ Test list
                ~ Test list
                ~~~ Test list
            ",
            "~~~not list",
            "~~> not a list",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn ordered_lists_regressions() {
        [
            "~ ~ a list item",
            "~~~~ - a list item",
            "~~~~ > a list item",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn quotes() {
        let examples: Vec<_> = [
            "> Test quote",
            ">>>> Test quote",
            "
                > Test quote
                > Test quote
                >> Test quote
                >> Test quote
                > Test quote
                >>> Test quote
            ",
            ">>>not quote",
            // "> > a quote item",
            ">>- not a quote",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn quotes_regressions() {
        [
            "> > a list item",
            ">>>> - a list item",
            ">>>> ~ a list item",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn definitions() {
        let examples: Vec<_> = [
            "$ Term
               Definition",
            "$$ Term
                Long definition
             $$",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn definitions_regressions() {
        [
            "$ Term Definition",
            "$$ Term
                Long definition $$",
            "$$ Term
                Long definition
             $$text",
            "$$ Term
                Long definition
             $$ text",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn footnotes() {
        let examples: Vec<_> = [
            "^ Title
               Content",
            "^^ Title
                Long content
             ^^",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn footnotes_regressions() {
        [
            "^ Term Definition",
            "^^ Term
                Long definition ^^",
            "^^ Term
                Long definition
             ^^text",
            "^^ Term
                Long definition
             ^^ text",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn tables() {
        let examples: Vec<_> = [
            ": A1
               Cell content",
            ":: A1
                Long cell content.
             ::",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn tables_regressions() {
        [
            ": Term Definition",
            ":: Term
                Long definition ::",
            ":: Term
                Long definition
             ::text",
            ":: Term
                Long definition
             :: text",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .for_each(|str| {
            parse(&str).unwrap_err();
        });
    }

    #[test]
    fn infirm_tags() {
        let examples: Vec<_> = [
            ".tag",
            ".tag-name_with-complexchars",
            ".tag-name_ parameter",
            ".tag-name_ one\\ large\\ parameter",
            ".tag-name_ one\\ large\\ parameter &^@! third parameter",
            ".tag.name.image https://github.com/super-special/repo.git?text=hello&other_text=bye",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    proptest! {
        #[test]
        fn infirm_tags_proptests(tag_name in TAG_NAME_REGEX, parameter in TAG_PARAMETER_REGEX, multi_parameter in TAG_MULTI_PARAMETER_REGEX) {
            let tag = format!(".{} {} {}\n", tag_name, parameter, multi_parameter);

            // TODO: Ensure that the number of parameters parsed is correct?
            parse(&tag).unwrap();
        }
    }

    #[test]
    fn carryover_tags() {
        let examples: Vec<_> = [
            "+tag
             paragraph",
            "+tag-name_with-complexchars
             paragraph",
            "+tag-name_ parameter
             paragraph",
            "+tag-name_ one\\ large\\ parameter
             paragraph",
            "+tag-name_ one\\ large\\ parameter &^@! third parameter
             paragraph",
            "+tag.name.image https://github.com/super-special/repo.git?text=hello&other_text=bye
             paragraph",
            "#tag
             paragraph",
            "#tag-name_with-complexchars
             paragraph",
            "#tag-name_ parameter
             paragraph",
            "#tag-name_ one\\ large\\ parameter
             paragraph",
            "#tag-name_ one\\ large\\ parameter &^@! third parameter
             paragraph",
            "#tag.name.image https://github.com/super-special/repo.git?text=hello&other_text=bye
             paragraph",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    proptest! {
        #[test]
        fn carryover_tags_proptests(tag_name in TAG_NAME_REGEX, parameter in TAG_PARAMETER_REGEX, multi_parameter in TAG_MULTI_PARAMETER_REGEX) {
            let content = format!("#{} {} {}\nhello world!", tag_name, parameter, multi_parameter);

            parse(&content).unwrap();
        }
    }

    #[test]
    fn ranged_verbatim_tags() {
        let examples: Vec<_> = [
            r#"@code
               print("Hello world!")
               @end"#,
            r#"@code.some-text.here lua\ language second-parameter
               print("Hello world!")
               @end"#,
            r#"@some-complex_tag_ first-parameter #&*(&$!) third-parameter

               function hello()
                   print("Hello World")
               end

               hello()
               @end"#,
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    proptest! {
        #[test]
        // NOTE: `.*` may at some point generate an `@end` purely by chance. There is a basic
        // check against this, but this should probably be done as a filter in proptest.
        fn ranged_verbatim_tags_proptests(tag_name in TAG_NAME_REGEX, parameter in TAG_PARAMETER_REGEX, multi_parameter in TAG_MULTI_PARAMETER_REGEX, content in ".*") {
            if content.contains("@end") {
                return Ok(());
            }

            let content = format!("@{} {} {}\n{}\n@end", tag_name, parameter, multi_parameter, content);

            parse(&content).unwrap();
        }
    }

    #[test]
    fn ranged_tags() {
        let examples: Vec<_> = [
            r#"|example
               Hello world!
               |end"#,
            r#"|example.some-text.here one\ parameter second-parameter
                #carryover
                text within
               |end"#,
            r#"|some-complex_tag_ first-parameter #&*(&$!) third-parameter
                this is some text within
               |end"#,
            r#"|example
               * Hello world!
               |end"#,
            r#"|example
               |example
               * Hello world!
               |end
               |end"#,
            r#"=example
               Hello world!
               =end"#,
            r#"=example.some-text.here one\ parameter second-parameter
                #carryover
                text within
               =end"#,
            r#"=some-complex_tag_ first-parameter #&*(&$!) third-parameter
                this is some text within
               =end"#,
            r#"=example
               * Hello world!
               =end"#,
            r#"=example
               =example
               * Hello world!
               =end
               =end"#,
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    proptest! {
        #[test]
        // NOTE: `.*` may at some point generate an `@end` purely by chance. There is a basic
        // check against this, but this should probably be done as a filter in proptest.
        fn ranged_tags_proptests(tag_type in prop_oneof!["@", "|"], tag_name in TAG_NAME_REGEX, parameter in TAG_PARAMETER_REGEX, multi_parameter in TAG_MULTI_PARAMETER_REGEX, content in PARAGRAPH_REGEX) {
            if content.contains(format!("{}end", tag_type).as_str()) {
                return Ok(());
            }

            let content = format!("{tag_type}{tag_name} {parameter} {multi_parameter}\n{content}\n{tag_type}end");

            parse(&content).unwrap();
        }
    }

    #[test]
    fn paragraphs() {
        let examples: Vec<_> = [
            "hello, world!",
            "*hello, world!*",
            "*hello,
             world!*",
            "two

             paragraphs",
            "paragraph
             here

             another paragraph
             here.",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    proptest! {
        #[test]
        fn paragraphs_proptests(paragraph_content in PARAGRAPH_REGEX) {
            parse(&paragraph_content).unwrap();
        }
    }
}
