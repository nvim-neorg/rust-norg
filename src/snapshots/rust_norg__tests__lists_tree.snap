---
source: src/lib.rs
expression: examples
---
- - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 1
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: base
      content: []
- - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 1
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: one
      content:
        - NestableDetachedModifier:
            modifier_type: UnorderedList
            level: 2
            extensions: []
            text:
              Paragraph:
                - Token:
                    Text: two
            content: []
- - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 1
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: one
      content:
        - NestableDetachedModifier:
            modifier_type: UnorderedList
            level: 2
            extensions: []
            text:
              Paragraph:
                - Token:
                    Text: two
                - Token: Whitespace
                - Token:
                    Text: with
                - Token: Whitespace
                - Token:
                    Text: content
            content: []
        - NestableDetachedModifier:
            modifier_type: UnorderedList
            level: 2
            extensions: []
            text:
              Paragraph:
                - Token:
                    Text: two
                - Token: Whitespace
                - Token:
                    Special: (
                - Token:
                    Text: "2"
                - Token:
                    Special: )
            content:
              - NestableDetachedModifier:
                  modifier_type: UnorderedList
                  level: 3
                  extensions: []
                  text:
                    Paragraph:
                      - Token:
                          Text: three
                  content: []
  - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 1
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: one
      content: []
- - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 2
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: two
      content: []
  - NestableDetachedModifier:
      modifier_type: UnorderedList
      level: 1
      extensions: []
      text:
        Paragraph:
          - Token:
              Text: one
      content: []
